File: 2022/go/07.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type Day7 struct {
	root *directory
	cwd  *directory
}

func (*Day7) Day() int {
	return 7
}

func (*Day7) IsTest() bool {
	return false
}

func (puzzle *Day7) Run(input string) {
	for _, line := range strings.Split(input, "\n") {
		line = strings.TrimSpace(line)

		if len(line) > 0 {
			puzzle.parseLine(line)
		}
	}

	//puzzle.print()
	puzzle.part1()
	puzzle.part2()
}

const printIndentSize = 2

func (puzzle *Day7) part1() {
	total := 0

	puzzle.part1Core(puzzle.root, &total, 100000)

	fmt.Printf("Part1: %d\n", total)
}

func (puzzle *Day7) part1Core(dir *directory, total *int, threshold int) {
	if dir.totalSize <= threshold {
		*total += dir.totalSize
	}

	for _, sub := range dir.directories {
		puzzle.part1Core(sub, total, threshold)
	}
}

func (puzzle *Day7) part2() {
	const maxSize = 70_000_000
	const neededForUpdate = 30_000_000

	availableSize := maxSize - puzzle.root.totalSize
	needToFree := neededForUpdate - availableSize

	bestDirectory := puzzle.root
	puzzle.findDirectory(&bestDirectory, puzzle.root, needToFree)

	fmt.Printf("Part2: %d\n", bestDirectory.totalSize)
}

func (puzzle *Day7) findDirectory(best **directory, current *directory, size int) {
	for _, dir := range current.directories {
		if dir.totalSize >= size && dir.totalSize < (*best).totalSize {
			*best = dir
		}
		puzzle.findDirectory(best, dir, size)
	}
}

func (puzzle *Day7) print() {
	puzzle.root.print(0)
}

func (puzzle *Day7) parseCommand(line string) {
	if strings.HasPrefix(line, "cd ") {
		dir := line[3:]
		if strings.HasPrefix(dir, "/") {
			puzzle.root = puzzle.newDirectory("/", nil)
			puzzle.cwd = puzzle.root
		} else if dir == ".." {
			puzzle.cwd = puzzle.cwd.parent
		} else {
			puzzle.cwd = puzzle.cwd.ensureSubDirectory(puzzle, dir)
		}
	}

	// Nothing to do on 'ls', just wait for comming lines.
}

func (puzzle *Day7) parseFsEntry(line string) {
	if strings.HasPrefix(line, "dir ") {
		name := line[4:]
		puzzle.cwd.ensureSubDirectory(puzzle, name)
	} else {
		parts := strings.Split(line, " ")
		size, _ := strconv.Atoi(parts[0])
		name := parts[1]
		puzzle.cwd.addFile(puzzle, name, size)
	}
}

func (puzzle *Day7) parseLine(line string) {
	if strings.HasPrefix(line, "$ ") {
		puzzle.parseCommand(line[2:])
	} else {
		puzzle.parseFsEntry(line)
	}
}

type file struct {
	name      string
	size      int
	directory *directory

	_ struct{}
}

func (f *file) print(level int) {
	format := fmt.Sprintf("%%%ds- %%s (file, size=%%d)\n", level*printIndentSize)
	fmt.Printf(format, "", f.name, f.size)
}

type directory struct {
	name        string
	totalSize   int
	parent      *directory
	directories []*directory
	files       []*file

	_ struct{}
}

func (*Day7) newDirectory(name string, parent *directory) *directory {
	return &directory{
		name:        name,
		parent:      parent,
		directories: []*directory{},
		files:       []*file{},
	}
}

func (parent *directory) ensureSubDirectory(puzzle *Day7, name string) *directory {
	for _, existing := range parent.directories {
		if existing.name == name {
			return existing
		}
	}

	new := puzzle.newDirectory(name, parent)
	parent.directories = append(parent.directories, new)

	return new
}

func (dir *directory) updateTotalSize(deltaSize int) {
	dir.totalSize += deltaSize
	if dir.parent != nil {
		dir.parent.updateTotalSize(deltaSize)
	}
}

func (dir *directory) addFile(puzzle *Day7, name string, size int) {
	file := &file{
		name:      name,
		size:      size,
		directory: dir,
	}

	dir.files = append(dir.files, file)
	dir.updateTotalSize(size)
}

func (dir *directory) print(level int) {
	format := fmt.Sprintf("%%%ds- %%s (dir, totalSize=%%d)\n", level*printIndentSize)
	fmt.Printf(format, "", dir.name, dir.totalSize)

	for _, childDir := range dir.directories {
		childDir.print(level + 1)
	}

	for _, file := range dir.files {
		file.print(level + 1)
	}
}

File: 2022/cs/07.cs
namespace Day7Namespace;

public class Day7 : IPuzzle
{
    public int Day => 7;
    public bool IsTest => false;

    public void Run(string input)
    {
        ParseInput(input);

        Part1();
        Part2();
    }

    private void Part1()
    {
        Console.WriteLine($"Part1: {Part1Core(root)}");
    }

    private int Part1Core(MyDirectory dir)
    {
        int result = dir.TotalSize <= 100_000 ? dir.TotalSize : 0;

        foreach (MyDirectory sub in dir.Directories)
            result += Part1Core(sub);

        return result;
    }

    private void Part2()
    {
        const int maxSize = 70_000_000;
        const int neededForUpdate = 30_000_000;

        int availableSize = maxSize - root.TotalSize;
        int needToFree = neededForUpdate - availableSize;

        MyDirectory bestDirectory = root;
        FindDirectory(ref bestDirectory, root, needToFree);

        Console.WriteLine($"Part2: {bestDirectory.TotalSize}");
    }

    private void FindDirectory(ref MyDirectory best, MyDirectory current, int size)
    {
        foreach (MyDirectory dir in current.Directories)
        {
            if (dir.TotalSize >= size && dir.TotalSize < best.TotalSize)
                best = dir;
            FindDirectory(ref best, dir, size);
        }
    }

    private MyDirectory root = new MyDirectory("/", null);
    private MyDirectory? cwd;

    private void ParseInput(string input)
    {
        cwd = root;

        foreach (string line in input.Split('\n').Select(x => x.Trim()))
        {
            if (string.IsNullOrWhiteSpace(line))
                continue;

            if (line.StartsWith("$ "))
                ParseCommand(line.Substring(2));
            else
                ParseFsEntry(line);
        }
    }

    private void ParseCommand(string command)
    {
        string[] parts = command.Split(' ');

        if (parts[0] == "cd")
        {
            string dir = parts[1];

            if (dir == "/")
                cwd = root;
            else if (dir == "..")
                cwd = cwd!.Parent;
            else
                cwd = cwd!.EnsureSubDirectory(dir);
        }

        // Nothing to do for ls command.
    }

    private void ParseFsEntry(string line)
    {
        string[] parts = line.Split(' ');

        if (parts[0] == "dir")
            cwd!.EnsureSubDirectory(parts[1]);
        else
        {
            int size = int.Parse(parts[0]);
            cwd!.Files.Add(new MyFile(parts[1], size, cwd!));
        }
    }
}

public class MyDirectory
{
    public string Name { get; }
    public MyDirectory? Parent { get; }
    public int TotalSize { get; private set; }
    public List<MyDirectory> Directories { get; } = new();
    public List<MyFile> Files { get; } = new();

    public MyDirectory(string name, MyDirectory? parent)
    {
        Name = name;
        Parent = parent;
    }

    public MyDirectory EnsureSubDirectory(string name)
    {
        MyDirectory? existing = Directories.Find(x => x.Name == name);

        if (existing != null)
            return existing;

        MyDirectory newDir = new(name, this);
        Directories.Add(newDir);

        return newDir;
    }

    public void UpdateTotalSize(int deltaSize)
    {
        TotalSize += deltaSize;
        if (Parent != null)
            Parent.UpdateTotalSize(deltaSize);
    }
}

public class MyFile
{
    public MyDirectory Parent { get; }
    public string Name { get; }
    public int Size { get; }

    public MyFile(string name, int size, MyDirectory parent)
    {
        Name = name;
        Size = size;
        Parent = parent;

        Parent.UpdateTotalSize(size);
    }
}
